llarp/dns.cpp

```diff
@@ -124,10 +124,180 @@ code_domain(char *&buffer, const std::string &domain) throw()
   *buffer++ = 0;
 }
 
-// lets just remove uint
-#ifdef _WIN32
-#define uint UINT
-#endif
+void
+vcode_domain(std::vector< byte_t > &bytes, const std::string &domain) throw()
+{
+  std::string::size_type start(0);
+  std::string::size_type end;  // indexes
+  // llarp::LogInfo("domain [", domain, "]");
+  while((end = domain.find('.', start)) != std::string::npos)
+  {
+    bytes.push_back(end - start);  // label length octet
+    for(std::string::size_type i = start; i < end; i++)
+    {
+      bytes.push_back(domain[i]);  // label octets
+      // llarp::LogInfo("Writing ", domain[i], " at ", i);
+    }
+    start = end + 1;  // Skip '.'
+  }
+
+  // llarp::LogInfo("start ", start, " domain size ", domain.size());
+
+  bytes.push_back(domain.size() - start);  // last label length octet
+  for(size_t i = start; i < domain.size(); i++)
+  {
+    bytes.push_back(domain[i]);  // last label octets
+    // llarp::LogInfo("Writing ", domain[i], " at ", i);
+  }
+  bytes.push_back(0);  // end it
+}
+
+// expects host order
+void
+vput16bits(std::vector< byte_t > &bytes, uint16_t value) throw()
+{
+  char buf[2]        = {0};
+  char *write_buffer = buf;
+  htobe16buf(write_buffer, value);
+  bytes.push_back(buf[0]);
+  bytes.push_back(buf[1]);
+}
+
+// expects host order
+void
+vput32bits(std::vector< byte_t > &bytes, uint32_t value) throw()
+{
+  char buf[4]        = {0};
+  char *write_buffer = buf;
+  htobe32buf(write_buffer, value);
+  bytes.push_back(buf[0]);
+  bytes.push_back(buf[1]);
+  bytes.push_back(buf[2]);
+  bytes.push_back(buf[3]);
+}
+
+void
+dns_writeType(std::vector< byte_t > &bytes, llarp::dns::record *record)
+{
+  llarp::dns::type_1a *type1a = dynamic_cast< llarp::dns::type_1a * >(record);
+  if(type1a)
+  {
+    std::vector< byte_t > more_bytes = type1a->to_bytes();
+    llarp::LogDebug("[1]Adding ", more_bytes.size());
+    bytes.insert(bytes.end(), more_bytes.begin(), more_bytes.end());
```

Comment:
Should a call to `bytes.reserve(bytes.size() + more_bytes.size())` be made before this insert?

Reply:
I don't see how that would be better, insert() has all the information to do it itself. If I was looping by char, I think it would make sense.